flink/v1.12/flink/src/main/java/com/netease/arctic/flink/catalog/ArcticCatalog.java

```diff
@@ -324,8 +334,40 @@ public void alterTable(ObjectPath tablePath, CatalogBaseTable newTable, boolean
   }
 
   @Override
-  public List<CatalogPartitionSpec> listPartitions(ObjectPath tablePath) throws CatalogException {
-    throw new UnsupportedOperationException();
+  public List<CatalogPartitionSpec> listPartitions(ObjectPath tablePath)
+      throws CatalogException, TableNotPartitionedException {
+    TableIdentifier tableIdentifier = getTableIdentifier(tablePath);
+    ArcticTable arcticTable = internalCatalog.loadTable(tableIdentifier);
+
+    if (arcticTable.spec().isUnpartitioned()) {
+      throw new TableNotPartitionedException(internalCatalog.name(), tablePath);
+    }
+    Set<CatalogPartitionSpec> set = Sets.newHashSet();
+    List<Table> tables = new LinkedList<>();
+
+    if (arcticTable.isKeyedTable()) {
+      tables.add(arcticTable.asKeyedTable().changeTable());
+      tables.add(arcticTable.asKeyedTable().baseTable());
+    } else {
+      tables.add(arcticTable.asUnkeyedTable());
+    }
+    for (Table table : tables) {
+      try (CloseableIterable<FileScanTask> tasks = table.newScan().planFiles()) {
```

Comment:
Mixed table Files is not simple change files + base files. you can use mixed plan

Reply:
I had talked with the auth of ServerTableDescriptor. The logic is error.

Reply:
> I had talked with the auth of ServerTableDescriptor. The logic is error.

Where can I get some reference about this logic?

Reply:
You can use KeyedTable#newScan to plan files

Reply:
> You can use KeyedTable#newScan to plan files

It seems cannot get partition or file information through this.

![image](https://github.com/NetEase/amoro/assets/40817998/907c2df8-689a-4ad5-8cd4-d298cc04b085)
There is no `planfiles` method,  and `plantask` cannot get file and partition information.

Reply:
planTasks -> CombinedScanTask -> KeyedTableScanTask -> ArcticFileScanTask -> PrimaryKeyedFile

Reply:
planTasks() -> KeyedTableScanTask -> ArcticFileScanTask. ArcticFileScanTask is a FileScanTask.

Reply:
> planTasks() -> KeyedTableScanTask -> ArcticFileScanTask. ArcticFileScanTask is a FileScanTask. or planTasks() -> KeyedTableScanTask -> ArcticFileScanTask -> PrimaryKeyedFile. PrimaryKeyedFile is a ContentFile

done